### ISSUE TITLE
Critical Bug - Sessions

### ISSUE THREAD

**Issue by evensis** [nexcess/magento-turpentine#806]
Hi there,

We've been having slow response times on our servers for quite some time now, and we've noticed that the issue was introduced as part of the 1.9.1.0 upgrade. We've now narrowed down the issue after a few email campaigns, which newrelic showed ESI being at fault.

For some reason, ESI goes nuts with session creation when loading up a page which has a few ESI requests in. I have applied jharrison's fix for this, however this didn't resolve it.

The one major issue this causes, is Varnish is unable to handle substantial concurrency. If we have over 100 concurrent visitors on the website, load shot up through the roof (we were seeing loads of 180+). This was due to session locks being in place, and each ESI request stacking up behind each other waiting for the session lock to be released.

This is a major bug, when we disabled varnish and reverted back to nginx and a custom page caching module I wrote, we were then able to handle 300 concurrent visitors with an average load of 1. With Varnish enabled, anything over 100 concurrent users would result in a sky rocketing load situation which brought the site to near standstill.

I think it's advisable that ESI shouldn't be able to do any form of writing to sessions, reading is fine, but the writing is what is causing the problem. Bot's were also making the situation significantly worse, in that for each bot request a session was opened along with ESI sessions being created as well.

For example, with Varnish enabled, we would have around 17,000 session keys open at any one time. With it now switched off, we have 800 keys open at any one time, which is a marked difference. The website is now responding sub 400ms, whereas with Varnish enabled it was hovering around 1500ms response.

Based on this bug, we have now disabled Varnish indefinitely. In all fairness, if using 1.9.1.0 and Turpentine, it shouldn't be used in a production environment where high concurrency is expected.

Thanks

**Comment by aricwatson** [comment]
Thanks for the report. Let me poke some people here to see what they can report about this. Just a thought -  did you try using something like memcache for sessions? We generally recommend that, especially if the site is running on a cluster. It sounds like you're using file-based sessions which may be a part of the problem.

**Comment by ichalov** [comment]
You may want to extend the list of bots ('Crawler User Agents' setting in 'Caching Options' page). Most of the bots are session-less, they don't supply the required cookie, so varnish tries to create a new session on each request from them. Bingbot is the most annoying of this kind and it's not included in the standard list. I believe it's not about migrating to 1.9.1.0 because I had problems with turpentine implementation increasing sessions creation in the earlier versions of magento too.

**Comment by aricwatson** [comment]
Yes, and the latest version (0.6.2) includes a fix for the unecessary session generation, so it's definitely worth trying that. It's from this PR https://github.com/nexcess/magento-turpentine/pull/516 if you wish to only make that change.

**Comment by evensis** [comment]
I had already applied #516, we're also using Redis for sessions rather than memcached. I had updated the list of bots substantially actually, bingbot being one of the ones we identified. We didn't have this issue in 1.6.2.0, and the site happily handled 600 concurrent users, whereas with 1.9.1.0 anything over 100 resulted in crippling performance issues.

**Comment by ichalov** [comment]
@evensis , do you remember what versions of varnish and turpentine were used with your magento 1.6.2.0 installation? I can confirm that session-less visit creates session records for every ESI block on the page with varnish 3.0.7 and these combinations: magento 1.6.2.0 + turpentine 0.6.1, magento 1.9.1.1 + turpentine 0.6.1, magento 1.9.1.1 + turpentine current master.

a3b4c3b421dfd01cf93b2081bd79bb6aaa7c82c2 doesn't fix it (it still leaves ESI requests without frontend cookie so magento creates new sessions for each of them).

**Comment by aricwatson** [comment]
Sorry for the long silence on my part, was on vacation (yay) - I did however check with our team that handles setting up Turpentine on larger sites and they haven't ever seen performance issues like this. I know that doesn't help much :(

> We didn't have this issue in 1.6.2.0, and the site happily handled 600 concurrent users, whereas with 1.9.1.0 anything over 100 resulted in crippling performance issues.

This makes sense, as the session key creating code was added to deal with the CRSF protection added in Magento 1.8. I just did some testing and I am seeing multiple sessions being created in vanilla install when I go directly to a product page. I'll poke around a little more.

**Comment by evensis** [comment]
Hi @aricwatson, @ichalov 

It was version 0.6.0 with 1.6.2.0 if I remember rightly, we updated to 0.6.1 (for 1.9.1.0) for the form key handling.

Varnish version was: varnishd (varnish-3.0.5 revision 1a89b1f)

Thanks

**Comment by evensis** [comment]
Nice to see it fixed @aricwatson, although the reason you didn't see this on other high concurrency environments is below.

So just to update, we've now managed to narrow down an issue we believe was compounded by the session issue. GFS2 (used for active/active VM clustered environments) partition has a nasty habit of deadlocking files, we believe that ESI session requests polling CM_RedisSession class files due to our APC configuration, and  the shear number of session requests grabbing at the class files, resulted in deadlocking the session files. So when we experienced a ton of requests due to a mail campaign for example, the additional users and the creation and polling of session class files, while a bug still, just compounded an issue on our environment causing the high load.

So in summary, this would only cause a high load situation on an environment where the GFS2 partition type is in use and apc_stat = 1 or worse no opcode cache is configured.

**Comment by aricwatson** [comment]
Very interesting, thank you @evensis - I'll pass this info on to our team that handles high-traffic Magento sites in case it's useful.
